File: 02_구현_기출문제/03_문자열_압축.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(string s) {
    int len = s.length();
    int answer = len;

    // len/2 보다 큰 단위로 자르지 못하기 때문에 len/2까지만 실행
    for (int unit = 1; unit <= len/2; unit++)
    {
        string final_str = "";
        string cur_unit_str = s.substr(0, unit); // 0부터 0+unit 직전까지 자르기 -> 제일 앞부터 정해진 길이만큼 잘라야 한다는 문제 조건이 있기 때문에 
        int cnt = 1; // 반복되는 unit 개수 

        for (int i = unit; i < len; i += unit)
        {
            string next_unit_str = s.substr(i, unit);
            // unit=2 기준, s[0]s[1]과 s[2]s[3]이 같다면
            if(cur_unit_str == next_unit_str) {
                cnt += 1;
            }
            // unit=2 기준, s[0]s[1]과 s[2]s[3]이 다르다면 
            else {
                // cur_unit이 반복된 개수가 1개 이하이면 1cur_unit 으로 하면 더 길어지니까 2이상인 경우에만 2cur_unit으로 한다.
                final_str += (cnt >= 2) ? to_string(cnt)+cur_unit_str : cur_unit_str;
                cur_unit_str = next_unit_str;
                cnt = 1;
            }
        }
        // for문에서 처리되지 못하고 남은 문자들을 처리해준다.
        final_str += (cnt >= 2) ? to_string(cnt)+cur_unit_str : cur_unit_str;
        answer = (answer < final_str.length() ? answer : final_str.length());
    }
    
    return answer;
}

int main(void){

    cout << solution("aabbaccc") << endl;
    cout << solution("xababcdcdababcdcd") << endl;

    return 0;
}

/*

abcabcabcabcdededededede
-> 14

xababcdcdababcdcd
17

*/

File: 02_구현_기출문제/03_문자열_압축.py
def solution(s):
    answer = len(s)
    # 1개 단위(unit)부터 압축 단위를 늘려가며 확인
    for unit in range(1, len(s) // 2 + 1):
        finalStr = ""
        curStr = s[0:unit] # 앞에서부터 unit만큼의 문자열 추출
        count = 1
        # 단위(unit) 크기만큼 증가시키며 이전 문자열과 비교
        for j in range(unit, len(s), unit):
            # 이전 상태와 동일하다면 압축 횟수(count) 증가
            nextStr = s[j:j + unit]
            if curStr == nextStr:
                count += 1
            # 다른 문자열이 나왔다면(더 이상 압축하지 못하는 경우라면)
            else:
                finalStr += str(count) + curStr if count >= 2 else curStr
                curStr = nextStr # 다시 상태 초기화
                count = 1
        # 남아있는 문자열에 대해서 처리
        finalStr += str(count) + curStr if count >= 2 else curStr
        # 만들어지는 압축 문자열이 가장 짧은 것이 정답
        print(finalStr)
        answer = min(answer, len(finalStr))
    return answer

print(solution("abcabcabcabcdededededede"))